libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxSpeechLoader.kt

```diff
@@ -17,28 +20,93 @@ import com.mapbox.navigation.ui.utils.internal.resource.ResourceLoader
 import com.mapbox.navigation.ui.utils.internal.resource.load
 import com.mapbox.navigation.ui.voice.model.TypeAndAnnouncement
 import com.mapbox.navigation.ui.voice.options.MapboxSpeechApiOptions
+import com.mapbox.navigation.utils.internal.InternalJobControlFactory
+import com.mapbox.navigation.utils.internal.logE
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.suspendCancellableCoroutine
 import java.net.MalformedURLException
 import java.net.URL
+import kotlin.coroutines.resume
 
-internal class MapboxSpeechProvider(
+internal class MapboxSpeechLoader(
     private val accessToken: String,
     private val language: String,
     private val urlSkuTokenProvider: UrlSkuTokenProvider,
     private val options: MapboxSpeechApiOptions,
-    private val resourceLoader: ResourceLoader
+    private val resourceLoader: ResourceLoader,
 ) {
 
-    suspend fun load(typeAndAnnouncement: TypeAndAnnouncement): Expected<Throwable, ByteArray> {
+    private val downloadedInstructions = mutableSetOf<TypeAndAnnouncement>()
+    private val downloadedInstructionsLock = Any()
+    private val defaultScope = InternalJobControlFactory.createDefaultScopeJobControl().scope
+
+    suspend fun load(
+        voiceInstruction: VoiceInstructions,
+        onlyCache: Boolean
+    ): Expected<Throwable, ByteArray> {
         return runCatching {
-            val url = instructionUrl(typeAndAnnouncement.announcement, typeAndAnnouncement.type)
-            val response = resourceLoader.load(url)
+            val typeAndAnnouncement = VoiceInstructionsParser.parse(voiceInstruction)
+                .getValueOrElse { throw it }
+            val request = createRequest(typeAndAnnouncement).apply {
+                if (onlyCache) {
+                    networkRestriction = NetworkRestriction.DISALLOW_ALL
+                }
+            }
+            val response = resourceLoader.load(request)
```

Comment:
Why is this needed? 

It seems like the load should try to get it from cache, if it cannot get it - load it from the network. Why do we need to make the caller specify? It seems like it will only create more errors

Reply:
> Why do we need to make the caller specify?

Sorry, I didn't quite get what you were asking.

> It seems like the load should try to get it from cache, if it cannot get it - load it from the network.

Well, the idea here is that we can specify whether we want to load id from the network if it's not in the cache or we don't. We'll use `false` when we don't have predownloading enabled and `true` otherwise.

Reply:
It looks like it would work without this logic, and it is common behavior to check cache and then only request from network on cache fail. So I'm confused why the caller should care if it is loaded from network or cache.

Reply:
The caller should care because only the caller knows if predownloading is enabled.
That's the current paradigm for low-level API users. If they want predownloading, they invoke a different method when the instruction should be played (`MaboxSpeechAPI#generatePredownloaded` instead of `MaboxSpeechAPI#generate`). If there was only one method for all cases that looks in the cache and if it's not present tries to download, we would still have the original issue with instruction being played too late. The idea here is that for predownloading mode we only look in the cache and if there's no instruction, we use the TTS.

Reply:
I don't think we need a `generatePredownloaded` because `generate` would work if it had the ability load the instruction from cache before requesting it from network. The low level api that is being added, is an ability to predownload instructions and to retrieve the predownloaded instructions. the generate function would not change for old versions because they were never predownloaded. does that make sense?

For example, add a function that can prefetch the `VoiceInstructions`. The caller determines how many voice instructions to prefetch. 

``` kotlin
val voiceInstructions = directionsRoute.legs()
            ?.flatMap { legs ->
                legs.steps()?.flatMap { steps -> steps.voiceInstructions() ?: emptyList() }
                    ?: emptyList()
            }
mapboxSpeechApi.prefetch(voiceInstructions)

// After the network requests are complete, this generate call will not request from network
// To remove issue where generate would double fetch, it would need manage "request handles". 
mapboxSpeechApi.generate(voiceInstructions.first())
```

Reply:
> the generate function would not change for old versions because they were never predownloaded.

And it didn't change. `generate` method still tries to get the instruction from cache and, if it fails, downloads it.
But I added `generatePredownloaded` and the difference here is that it won't start a network request. I mean what if predownloading failed? Or didn't have enough time to finish? We don't want to start a network request in this case because it will delay playing the instruction until the request finishes. This behaviour used to cause the bug with the instruction being played too late. So we want to eliminate this possibility.

Reply:
How many times will it really fail though, and how many times will there really be double requests? I'm just proposing an alternative approach to this, because I also do not think it will eliminate the issue - we're just reducing it.

Reply:
Actually, I'd suggest pausing this discussion for now because if I switch to using our own files as cache as I described in Slack, it will be irrelevant. If not, we'll get back to it.

Reply:
I've added a separate commit to address the caching issue. The `onlyCache` parameter is gone.
Do you have any other concerns here?

Reply:
Thanks for digging and addressing! Resolving here